iml-services/iml-snapshot/src/retention.rs

```diff
@@ -0,0 +1,213 @@
+    let resp: iml_graphql_queries::Response<iml_graphql_queries::snapshot::destroy::Resp> =
+        graphql(
+            client,
+            iml_graphql_queries::snapshot::destroy::build(fs_name, snapshot_name, true),
+        )
+        .await?;
+
+    let cmd = Result::from(resp)?.data.destroy_snapshot;
+
+    Ok(cmd)
+}
+
+async fn get_retention_filesystems(pool: &PgPool) -> Result<Vec<String>, Error> {
+    let xs = get_retentions(pool).await?;
+
+    let fs_names = xs.into_iter().map(|x| x.filesystem_name).collect();
+    Ok(fs_names)
+}
+
+pub async fn process_retention(
+    client: &Client,
+    influx_client: &InfluxClient,
+    pool: &PgPool,
+    mut stats_record: HashMap<String, u64>,
+) -> Result<HashMap<String, u64>, Error> {
+    let filesystems = get_retention_filesystems(pool).await?;
+
+    tracing::debug!("Filesystems with retentions: {:?}", filesystems);
+
+    for fs_name in filesystems {
+        let stats = get_stats_from_influx(&fs_name, &influx_client).await?;
+
+        if let Some((bytes_total, bytes_free, bytes_used)) = stats {
+            tracing::debug!(
+                "stats values: {}, {}, {}",
+                bytes_total,
+                bytes_free,
+                bytes_used
+            );
+            let percent_used = (bytes_used as f64 / bytes_total as f64) as f64 * 100.0f64;
+            let percent_free = 100.0f64 - percent_used;
+
+            tracing::debug!(
+                "stats record: {:?} - bytes free: {}",
+                stats_record.get(&fs_name),
+                bytes_free
+            );
+            let retention = get_retention_policy(pool, &fs_name).await?;
+
+            if let Some(retention) = retention {
+                let snapshots = get_snapshots(pool, &fs_name).await?;
+
+                tracing::debug!(
+                    "percent_left: {}, reserve value: {}",
+                    percent_free,
+                    retention.reserve_value
+                );
+                let should_delete_snapshot = match retention.reserve_unit {
+                    snapshot::ReserveUnit::Percent => percent_free < retention.reserve_value as f64,
+                    snapshot::ReserveUnit::Gibibytes => {
+                        let gib_free: f64 = bytes_free as f64 / 1_073_741_824_f64;
+                        gib_free < retention.reserve_value as f64
+                    }
+                    snapshot::ReserveUnit::Tebibytes => {
+                        let teb_free: f64 = bytes_free as f64 / 1_099_511_627_776_f64;
+                        teb_free < retention.reserve_value as f64
+                    }
+                };
+                tracing::debug!("Should delete snapshot?: {}", should_delete_snapshot);
+
+                if should_delete_snapshot
+                    && snapshots.len() > retention.keep_num as usize
+                    && (stats_record.get(&fs_name).is_none()
+                        || stats_record.get(&fs_name) != Some(&bytes_used))
```

Comment:
I think `None != Some(&bytes_used)` :)

Reply:
You're right. I'm still used to having to do checks like this... I'll update.